Add EIP: Notification Interface (#7017)
ethereum/EIPs: EIPS/eip-7017.md

---
eip: 7017
title: Notification Interface
description: A notifications interface for a more engaging blockchain
author: Oliver Stehr (@Oli-art)
discussions-to: https://ethereum-magicians.org/t/eip-7017-notifications-interface-for-a-more-engaging-blockchain/11091
status: Draft
type: Standards Track
category: ERC
created: 2022-11-10
requires: 681, 5630
---

## Abstract

The following standard allows addresses / smart contracts to send notifications one-to-one and one-to-many.

It achieves it in a trustless and decentralized way and without requiring any on-chain interaction by the receivers.

It requires a front-end implementation where the notifications are listened to, filtered and showed to the end user (address owner).

Some usecases include but are not limited to:

- DAO governance voting anouncement
- DEX informing an address about a certain price limit being reached, for example a stop-loss or a margin-call.
- An NFT marketplace informing an NFT owner about an offer being made to one of it’s NFTs.
- A metaverse informing about an important event.
- Warning to an address about an ENS domain expiration date approaching.

## Motivation

> With the adoption of web3 applications, an increasing necessity arises to be informed about certain events happening on-chain.

Reviewer: How about this wording instead:

```suggestion
As the adoption of web3 applications grows, so too does the need to be informed about certain events happening on-chain.
```

> Users are used to being informed, whether it be about news or updates of their favorite applications. As we are in a time of instant data feeds on social media, instant messaging apps and notifications of events that users care about, notifications are a feature in practically every application that exists in web2. They mostly come to email inboxes, SMSs, inside the applications, or in the notification inbox in the operating system.

Reviewer: ```suggestion
Users are accustomed to being informed, whether it be about news or updates of their favorite applications. As we are in a time of instant data feeds on social media, instant messaging apps and notifications of events that users care about, notifications are a feature in practically every application that exists in web2. They mostly come to email inboxes, SMSs, inside the applications, or in the notification inbox in the operating system.
```

If they would be taken away, the engagement on these web2 applications would sink. This is not different with web3 applications: users cannot be left in the dark about what is going on in them. Not only that, for some applications, all that matters is the participation of users on certain events, like governance on a DAO.

In web2, most of the user account’s are linked to an email address that is required at sign-up. This makes it easy to send notifications to specific users and requires no further complexity as an infrastructure exists to deliver a message to the user using their email address.

In web3, on the other hand, there's mostly is only one inbox to send notifications to: addresses.

> Every smart contract can define its own event’s to which one can listen to, but for each of them, a change has to be done in the frontend to listen to that specific contract and event structure. This poses a problem of coordination between smart contracts and web3 applications that can notify users.

Reviewer: ```suggestion
Every smart contract can define its own events to which one can listen to, but for each of them, a change has to be done in the frontend to listen to that specific contract and event structure. This poses a problem of coordination between smart contracts and web3 applications that can notify users.
```

This EIP aims at proposing a decentralized approach to send and receive notifications from and to ethereum addresses, including smart contracts, in a standarized way, facilitating front-end intrgrations. These could be:

- Wallets
- Mobile phone dapps
- Email services
- Discord / Telegram Bots

All of these could be notifying about on-chain notifications to the user.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> An approach that is both simple, decentralized and easy to implement is to use a notifications smart contract standard that is **event-based** to be able to emit notifications to one address or broadcast them to anyone that wants to listen. Off-chain **whitelists** would record all addresses a user wants to allow receiving messages from to avoid spam. This is useful for direct messages from one address to another. Off-chain **subscription lists** would indicate which addresses they want to listen for general broadcasts. This is useful for receiving updates on a project.

Reviewer: This paragraph introduces no requirements or recommendations for implementation. It should be split and moved to motivation and rationale as appropriate.

> As a user SHALL NOT record its whitelist and subscription list on-chain, the receiver will not do any transaction.

Reviewer: I think this is a bit more readable:

```suggestion
Notification recipients SHALL NOT be required to record a whitelist and subscription list on-chain.
```

    */
    function walletBroadcast (string memory data) external virtual {
        emit Message (msg.sender, address(0), data);
    }

    /**
    * @dev Send a general notification from the address executing the function
    * @param data to broadcast: (json format, URI recomended):
    * {subject, body, attention_level (1, 2 or 3), image_url (optional),
    * transaction_request (optional)}
    */
    function contractBroadcast (string memory data) external virtual {
        emit Message (address(this), address(0), data);
    }
}
```

## Security Considerations

There are security considerations for the proposal regarding impersonation threats.

> It is up to the client-side application to listen only to contracts which make "sensible" choices as far as the meanings of from and to.

This is related to how the clients will interpret and filter the events. There can be malicious contracts that implement the notification interface but whose functions emit events where the sender is a third party that did not consent to send the message.

Here is an example of a malitious implementation of a BroadcastMsg event impersonating Vitalik Buterin:

```solidity

> emit BroadcastMsg (0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045, message);

Reviewer: This event doesn't match one from the interface. Is that intentional?

Also, every `walletDM` and `contractDM` function **MUST** implement the `DirectMsg` event with `from` set as `msg.sender` and `address(this)` respectively.

Here is a table to better represent this:

|          | DirectMessage                                     | Broadcast                                        |
| -------- | ------------------------------------------------- | ------------------------------------------------ |
| Wallet   | emit DirectMsg(msg.sender, to, message);    | emit BroadcastMsg(msg.sender, message);    |
| Contract | emit DirectMsg(address(this), to, message); | emit BroadcastMsg(address(this), message); |

### Data Schema

The `data` field **REQUIRED** in the message events **MUST** have a JSON schema that can be written directly as a string, or passed as a URI containing the JSON. It is **RECOMENDED** to use a URI, as it is a cheaper way to send the data.

It has two different specifications, one for broadcast messages and another one for direct messages. Only small differences exist between them. This is the base specifications for both:

```json
{
"subject": "The subject of the message",
"body": "The body of the message",
"image_uri": "An image URI",

> "transaction_request": "A transaction request"

Reviewer 2: it looks like you are missing indentations here. would recommend trying to match lines 151–154, and also clarifying how indentation/spacing should be handled in the JSONs.

    */
    function walletBroadcast (string memory data) external virtual override {
        emit BroadcastMsg (msg.sender, data);
    }

    /**
    * @dev Send a general notification from the address executing the function
    * @param data to broadcast: (json format, URI recomended):
    * {subject, body, attention_level (1, 2 or 3), image_url (optional),
    * transaction_request (optional)}
    */
    function contractBroadcast (string memory data) external virtual override {
        emit BroadcastMsg (address(this), data);
    }
}
```

## Security Considerations

### Impersonation

There are security considerations for the proposal regarding threats.

The first is related to how the clients will interpret and filter the events. There can be malicious contracts that implement the notification interface but whose functions emit events where the sender is neither the message sender nor the contract. They could try to impersonate another address by changing the sender.

Here is an example of a malitious implementation of a BroadcastMsg event impersonating Vitalik Buterin:

```solidity
emit BroadcastMsg (0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045, message);
```

> This is why the message sender in the log's topic must allways be verified against the transaction "from" or transaction "to", depending on the notification type (wallet or contract respectively).

Reviewer 2: of course, another method would be to have a "known" trusted deployed instance of this interface, which would be hardcoded in the client-side software. that way the client code could avoid performing this check explicitly, since it trusts the contract. this seems like the route that'd be taken in practice for messaging software.
